src/vm_pmu.rs

```diff
@@ -368,4 +375,167 @@ impl VmPmuState {
             }
         }
     }
+
+    /// Helper function to call either SBI, or the internal implementation
+    /// for configure_matching_counters() depending on whether Sscdeleg is implemented
+    pub fn configure_matching_counters(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::configure_matching_counters(
+                counter_index,
+                counter_mask,
+                config_flags,
+                event_type,
+                event_data,
+            )
+        } else {
+            fn is_firmware_counter(event_type: PmuEventType) -> bool {
+                matches!(event_type, PmuEventType::Firmware(_))
+            }
+
+            if is_firmware_counter(event_type) {
+                return Err(SbiError::InvalidParam);
+            }
+
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let index = if !config_flags.is_skip_match() {
+                // Fixed counters (0, 2) must use a static mapping pending finalization of event mapping
+                if let Some(index) = CounterMaskIter::new(counter_index, counter_mask)
+                    .into_iter()
+                    .next()
+                {
+                    if index <= 2 && !matches!(self.counter_state[index], NotConfigured) {
+                        return Err(SbiError::InvalidParam);
+                    }
+                }
```

Comment:
I'm not quite following this logic. Looks like we're just enforcing that if a fixed counter is in `counter_mask` that it be unconfigured, and thus will guaranteed to be picked below on line 419. I thought the plan was to always pick counter 0 if the event was `cycles` and counter 2 if the event was `instret`?

Reply:
As I see it, there's no canonical way of picking a counter because it's a combination of the starting counter index and the counter_mask. 

For example:

A counter_index of 0 and a mask of 0100b => find() returns counter 2
A counter index of 1 and a mask of 0010b => find() returns counter 2
A counter index of 2 and a mask of 0001b => find() returns counter 2
A counter index of 0 and a mask of xxx1b => find() returns counter 0

In each of the above cases, next() returns the same element as find(), and we enforce that it's not configured. @atishp04: Any comments on the above?

Reply:
Per offline conversation with @atishp04: First filter by event_type, and if CYCLES / INSRET, confirm that they are selected by the index / mask and are not configured. If OK, return a fixed counter, else error out early

Reply:
Also, skip_match takes precedence, so changed accordingly.